sys/kern/src/arch/arm_m.rs

```diff
@@ -1001,51 +906,61 @@ pub(crate) fn with_task_table<R>(
 /// pointer while you have access to `task`, and as long as the `task` being
 /// stored is actually in the task table, you'll be okay.
 pub unsafe fn set_current_task(task: &mut task::Task) {
-    // Safety: should be ok if the contract above is met (TODO: make me an
-    // atomic)
-    unsafe {
-        CURRENT_TASK_PTR = Some(NonNull::from(task));
-    }
+    CURRENT_TASK_PTR.store(task, Ordering::Relaxed);
 }
 
 /// Reads the tick counter.
 pub fn now() -> Timestamp {
-    Timestamp::from(unsafe { TICKS })
+    // Recall that we expect the systick interrupt cannot preempt kernel code,
+    // so we're safe to read this in two nonatomic parts here.
+    Timestamp::from([
+        TICKS[0].load(Ordering::Relaxed),
+        TICKS[1].load(Ordering::Relaxed),
+    ])
 }
 
 /// Kernel global for tracking the current timestamp, measured in ticks.
 ///
-/// This is a mutable `u64` instead of an `AtomicU64` because ARMv7-M doesn't
-/// have any 64-bit atomic operations. So, we access it carefully from
-/// non-preemptible contexts.
-static mut TICKS: u64 = 0;
+/// This is a pair of `AtomicU32` because (1) we want the interior mutability of
+/// the atomic types but (2) ARMv7-M doesn't have any 64-bit atomic operations.
+/// We access this only from contexts where we can't be preempted, so, the fact
+/// that it's split across two words is ok.
+///
+/// `TICKS[0]` is the least significant part, `TICKS[1]` the most significant.
+static TICKS: [AtomicU32; 2] = {
+    const ZERO: AtomicU32 = AtomicU32::new(0);
+    [ZERO; 2]
+};
 
 /// Handler that gets linked into the vector table for the System Tick Timer
 /// overflow interrupt. (Name is dictated by the `cortex_m` crate.)
 #[allow(non_snake_case)]
 #[no_mangle]
 pub unsafe extern "C" fn SysTick() {
     crate::profiling::event_timer_isr_enter();
-    // Safety: We configure this interrupt to have the same priority as SVC,
-    // which means there's no way this can preempt the kernel -- it will only
-    // preempt user code. As a result, we can manufacture exclusive references
-    // to various bits of kernel state.
-    let ticks = unsafe { &mut TICKS };
     with_task_table(|tasks| {
-        // Advance the kernel's notion of time.
-        // This increment is not expected to overflow in a working system, since it
-        // would indicate that 2^64 ticks have passed, and ticks are expected to be
-        // in the range of nanoseconds to milliseconds -- meaning over 500 years.
-        // However, we do not use wrapping add here because, if we _do_ overflow due
-        // to e.g. memory corruption, we'd rather panic and reboot than attempt to
-        // limp forward.
-        *ticks += 1;
-        // Now, give up mutable access to *ticks so there's no chance of a
-        // double-increment due to bugs below.
-        let now = Timestamp::from(*ticks);
-        drop(ticks);
+        // Load the time before this tick event.
```

Comment:
Is there any advantage to building the `Timestamp` outside of the closure and passing it in, to minimize how long the task table is held?

Reply:
Not really -- the task table in-use flag is not a mutex, and this is not a critical section (as the kernel is not preemptible). Moving the bounds of the closure would just rearrange instructions.